src/main/java/com/worksap/fig/lang/Range.java

```diff
@@ -0,0 +1,197 @@
+        Objects.requireNonNull(to);
+        this.to = to;
+        toIncluded = false;
+        return this;
+    }
+
+    /**
+     * Set next element generator.
+     * @param next The next element generator
+     * @return Current range object.
+     * @throws NullPointerException if next is null.
+     */
+    public Range<C> next(Function<C, C> next) {
+        Objects.requireNonNull(next);
+        this.next = next;
+        return this;
+    }
+
+    /**
+     * Set next element generator.
+     * @param next The next element generator
+     * @return Current range object.
+     * @throws NullPointerException if next is null.
+     */
+    public Range<C> next(BiFunction<C, Integer, C> next) {
+        Objects.requireNonNull(next);
+        this.biNext = next;
+        return this;
+    }
+
+    /**
+     * @return The start point
+     */
+    public C getFrom() {
+        return from;
+    }
+
+    /**
+     * @return The end point
+     */
+    public C getTo() {
+        return to;
+    }
+
+    /**
+     * @return Whether the end point is included in this range.
+     */
+    public boolean isToIncluded() {
+        return toIncluded;
+    }
+
+    /**
+     * Iterate each element of the range.
+     *
+     * @throws NullPointerException if action, this.from, this.to or this.next is null.
+     */
+    public void forEach(Consumer<? super C> action) {
+        forEach((e, i) -> action.accept(e));
+    }
+
+    /**
+     * Similar to {@link #forEach(Consumer)}, with additional parameter "index" as the second parameter of the lambda expression.
+     *
+     * @throws NullPointerException if action is null
+     */
+    public void forEach(BiConsumer<? super C, Integer> action) {
+        Objects.requireNonNull(action);
+        Objects.requireNonNull(from);
+        Objects.requireNonNull(to);
+
+        if (Objects.isNull(next) && Objects.isNull(biNext)) {
+            Objects.requireNonNull(next);
+        }
+
+        /* orientation = 0 means to is equal to from;
+         * orientation > 0 means to is greater than from;
+         * otherwise to is less than from.
+         */
+        int orientation = from.compareTo(to);
+        int i = 0;
+        C current = from;
+        int cmp = current.compareTo(to);
+        while (cmp * orientation > 0 || toIncluded && cmp == 0) {
+            action.accept(current, i);
+            if (Objects.nonNull(next)) {
+                current = next.apply(current);
+            } else {
+                current = biNext.apply(current, i);
+            }
+            Objects.requireNonNull(current);
+            cmp = current.compareTo(to);
+            ++i;
+        }
+    }
+
+    public Seq<C> toSeq() {
+        List<C> list = new ArrayList<>();
+        forEach(e -> list.add(e));
+        return Seqs.newSeq(list);
+    }
+
+    public MutableSeq<C> toMutableSeq() {
+        List<C> list = new ArrayList<>();
+        forEach(e -> list.add(e));
+        return Seqs.newMutableSeq(list);
```

Comment:
You can create a mutableSeq and use `appendInplace` to avoid a copy of the list.